Stateful 2fa providers

ChristophWurst opened the pull request:
This change makes two-factor providers and their association with users (whether the provider is enabled or not) *stateful* as in that we don't actually require the provider to be loaded when we want to check if a user uses 2FA. This greatly improves performance as we don't have to read from the file system and parse the `info.xml` file.

To do:
* [x] Basic implementation
* [x] Adapted/added tests
* [x] Squashed commits
* [ ] Adapt 2fa apps to use the `IRegistry` to propagate their enabled/disabled state
  + [ ] twofactor_u2f
  + [ ] twofactor_totp
  + [ ] twofactor_sms
  + [ ] others?
* ~~Proper handling of the backup codes app (auto-enable when other providers get enabled? auto-disable when all other providers get disabled?)~~ won't change that for now and keep the existing logic (app has to be configured by user).

cc @rullzer

ChristophWurst committed:
Make 2FA providers stateful

This adds persistence to the Nextcloud server 2FA logic so that the server
knows which 2FA providers are enabled for a specific user at any time, even
when the provider is not available.

The `IStatefulProvider` interface was added as tagging interface for providers
that are compatible with this new API.

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst commented:
This is ready for review now.

Steps to test:

* Enable a 2fa provider app
* Configure the 2fa provider
* Log in once
* Log out
* Disable the app again
* Try to log in

Before:
You could still log in.

Now:
You see a warning that at least one provider failed to load and you cannot complete the login.

ChristophWurst commented:
- [ ] failure 1

> There were 2 warnings:
> 
> 1) Tests\Core\Controller\LoginControllerTest::testLoginWithOneTwoFactorProvider
> Trying to configure method "getProviders" which cannot be configured because it does not exist, has not been specified, is final, or is static
> 
> 2) Tests\Core\Controller\LoginControllerTest::testLoginWithMultpleTwoFactorProviders
> Trying to configure method "getProviders" which cannot be configured because it does not exist, has not been specified, is final, or is static 

- [ ] failure 2

> 1) Test\Authentication\TwoFactorAuth\Db\ProviderUserAssignmentDaoTest::testPersist
> Failed asserting that 0 is identical to 1.
> 
> /drone/src/github.com/nextcloud/server/tests/lib/Authentication/TwoFactorAuth/Db/ProviderUserAssignmentDaoTest.php:93

ChristophWurst commented on tests/lib/Authentication/TwoFactorAuth/Db/ProviderUserAssignmentDaoTest.php:
```diff
@@ -0,0 +1,94 @@
+			->andWhere($qb->expr()->eq('enabled', $qb->createNamedParameter(0)));
+		$res = $q->execute();
+		$cnt = $res->rowCount();
+		$this->assertSame(1, $cnt);
```
For some weird reason the query returns `0` rows here. I cannot find out why that is the case. If I attach a debugger an step through the code I see that the `->persist` call inserts one row.

@rullzer did I use the query builder incorrectly?

rullzer replied:
Rowcount is the number of rows affected by insert/update/delete

Either do 'fetchAll' and count on that. Or use the count function in the query. Both should work just fine here.

ChristophWurst replied:
I should have read the method docs. Makes sense, updated my code and tests seem to pass locally. Thanks a lot!

MorrisJobke commented on version.php:
```diff
@@ -29,7 +29,7 @@
 // when updating major/minor version number.
 
-$OC_Version = array(14, 0, 0, 4);
+$OC_Version = array(14, 0, 0, 5);
```
Needs to be changed again, because another PR with the same change got merged.

ChristophWurst replied:
Yep https://github.com/nextcloud/server/pull/9485#discussion_r196305387

rullzer commented:
Very cool stuff. I need to test it properly but code makes a lot of sense

ChristophWurst commented:
Let's see if CI likes the current state and if so, I'll squash my commits one more time.

ChristophWurst commented on core/Migrations/Version14000Date20180522074438.php:
```diff
@@ -0,0 +1,82 @@
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 */
+	public function preSchemaChange(IOutput $output, Closure $schemaClosure,
```
done

ChristophWurst commented on core/Migrations/Version14000Date20180522074438.php:
```diff
@@ -0,0 +1,82 @@
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 */
+	public function postSchemaChange(IOutput $output, Closure $schemaClosure,
```
done

MorrisJobke commented:
> Let's see if CI likes the current state and if so, I'll squash my commits one more time.

And now? xD

ChristophWurst commented:
> And now? xD

Squashed, rebased and CI is happy. Time to review :)

Will fix the 2fa providers soonish so that they propagate their state automatically.

rullzer commented on lib/private/Authentication/TwoFactorAuth/ProviderLoader.php:
```diff
@@ -0,0 +1,88 @@
+<?php
+
```
strict

rullzer commented on lib/private/Authentication/TwoFactorAuth/ProviderLoader.php:
```diff
@@ -0,0 +1,88 @@
+	 * @param string $appId
+	 */
+	protected function loadTwoFactorApp(string $appId) {
+		if (!OC_App::isAppLoaded($appId)) {
```
Can't we use the appmanager here?

ChristophWurst replied:
Neither the interface, nor the implementation provide methods to load an app.

MorrisJobke replied:
See #8505

rullzer commented on lib/private/Authentication/TwoFactorAuth/ProviderSet.php:
```diff
@@ -0,0 +1,71 @@
+<?php
+
```
strict

rullzer reviewed (approved):
Cool stuff!!!

ChristophWurst committed:
Make new classes strict and fix license header

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

MorrisJobke commented on core/templates/twofactorselectchallenge.php:
```diff
@@ -1,6 +1,11 @@
 	<p><?php p($l->t('Enhanced security is enabled for your account. Please authenticate using a second factor.')) ?></p>
+	<?php if ($_['providerMissing']): ?>
+	<p>
+		<strong><?php p($l->t('Could not load at least one of your enabled two-factor auth methods. Please contact your admin.')) ?></strong>
```
Do we want to add an setup check for this as well? Just iterate over the table and list them for the admin?

ChristophWurst replied:
Like querying a set of generally enabled providers and checking if they can be loaded? Sounds reasonable.

ChristophWurst replied:
-> https://github.com/nextcloud/server/issues/9985

MorrisJobke reviewed (approved):
👍

MorrisJobke commented:
CI is here: https://drone.nextcloud.com/nextcloud/server/8327

ChristophWurst commented:
The one failing acceptance tests looks unrelated to this change.

nickvergessen commented on core/Command/TwoFactorAuth/State.php:
```diff
@@ -0,0 +1,110 @@
+
+namespace OC\Core\Command\TwoFactorAuth;
+
+use OC\Core\Command\Base;
```
PHP Fatal error: Cannot use OC\Core\Command\Base as Base because the name is already in use in /drone/src/github.com/nextcloud/server/core/Command/TwoFactorAuth/State.php on line 29

nickvergessen replied:
See https://github.com/nextcloud/server/pull/10023
